src/cart/crt_hg_proc.c

```diff
@@ -114,14 +115,103 @@ CRT_PROC_TYPE_FUNC(uint64_t)
 CRT_PROC_TYPE_FUNC(bool)
 
 int
-crt_proc_crt_bulk_t(crt_proc_t proc, crt_proc_op_t proc_op,
-		    crt_bulk_t *bulk_hdl)
+crt_proc_crt_bulk_t(crt_proc_t proc, crt_proc_op_t proc_op, crt_bulk_t *pcrt_bulk)
 {
-	hg_return_t	hg_ret;
+	struct crt_bulk *bulk = NULL;
+	hg_return_t      hg_ret;
+	hg_bulk_t        tmp_hg_bulk;
 
-	hg_ret = hg_proc_hg_bulk_t(proc, (hg_bulk_t *)bulk_hdl);
+	/*
+	 * We only send 'hg_bulk_t' over the wire. During encoding stage we
+	 * extract mercury bulk handle from crt_bulk_t, while during decode
+	 * we wrap mercury bulk in a newly allocated crt_bulk_t struct
+	 */
+	switch (proc_op) {
+	case CRT_PROC_ENCODE:
+		bulk = *pcrt_bulk;
 
-	return (hg_ret == HG_SUCCESS) ? 0 : -DER_HG;
+		/* RPC can have a NULL bulk. if so, encode a NULL value */
+		if (!bulk) {
+			tmp_hg_bulk = HG_BULK_NULL;
+			hg_ret      = hg_proc_hg_bulk_t(proc, (hg_bulk_t *)&tmp_hg_bulk);
+			return (hg_ret == HG_SUCCESS) ? 0 : -DER_HG;
+		}
+
+		/* Deferred allocation as a result of D_QUOTA_BULKS limit */
+		if (bulk->deferred) {
+			struct crt_context *ctx;
+			int                 rc;
+
+			/* Create mercury handle based on saved params */
+			ctx = bulk->crt_ctx;
+			D_ASSERT(ctx != NULL);
+
+			rc = crt_hg_bulk_create(&ctx->cc_hg_ctx, &bulk->sgl, bulk->bulk_perm,
+						&bulk->hg_bulk_hdl);
+			if (rc != DER_SUCCESS)
+				return rc;
+
+			record_quota_resource(ctx, CRT_QUOTA_BULKS);
+
+			if (bulk->bound) {
+				rc = crt_hg_bulk_bind(bulk->hg_bulk_hdl, &ctx->cc_hg_ctx);
+				if (rc != 0) {
+					D_ERROR("Failed to bind bulk during proc\n");
+					/* free will return quota resource */
+					crt_bulk_free(bulk->hg_bulk_hdl);
+					return rc;
+				}
+			}
+			bulk->deferred = false;
+		}
+
+		/* Send mercury bulk handle over the wire */
+		hg_ret = hg_proc_hg_bulk_t(proc, (hg_bulk_t *)&bulk->hg_bulk_hdl);
+		return (hg_ret == HG_SUCCESS) ? 0 : -DER_HG;
+		break;
+
+	case CRT_PROC_DECODE:
+		/* unpack mercury handle and wrap it around crt_bulk_t struct */
+		hg_ret = hg_proc_hg_bulk_t(proc, &tmp_hg_bulk);
+		if (hg_ret != HG_SUCCESS)
+			return -DER_HG;
+
+		/* don't create a bulk wrapper if null bulk was transmitted */
+		if (tmp_hg_bulk == HG_BULK_NULL) {
+			*pcrt_bulk = NULL;
+			return 0;
+		}
+
+		/* Allocate space for a wrapper struct */
+		D_ALLOC_PTR(bulk);
+		if (!bulk)
+			return -DER_NOMEM;
```

Comment:
since servers don't use quota and never use that struct I guess you could consider having a hack that skips the allocation there and only assigns the bulk handle but it's maybe a bit too hacky... just some thoughts... (similarly clients never call bulk_transfer)

Reply:
hmm not sure how that would work. we still use functions shared between client and server that expect crt_bulk_t and will extract hg_bulk_hdl from it. if engines only used 'hg_bulk_hdl' variant of functions then yeah i can see that.